### PR TITLE
fix #20675 by adding an esbuild plugin to inject env variables

### LINKED ISSUE #20675: [Bug]: sb7+vite env variables not working inside addons
### Describe the bug

Environment variables are not available inside addon.

This is a repo to reproduce the error: node16 + sb7 + vite + npm
https://github.com/Luk-z/vite_ts_sb7_npm

This is a repo where the env variables works inside addons: node14 + sb6 + vite + npm
https://github.com/Luk-z/vite_ts_sb6_npm

To write the simple addon i followed https://storybook.js.org/docs/react/addons/writing-addons.

I'm expecting env variables are available inside an addon.

### To Reproduce

```shell
https://github.com/Luk-z/vite_ts_sb7_npm
```

### System

```shell
$ npx sb@next info

Environment Info:

  System:
    OS: macOS 12.6
    CPU: (10) arm64 Apple M1 Pro
  Binaries:
    Node: 16.19.0 - ~/.nvm/versions/node/v16.19.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.12.1/bin/yarn
    npm: 8.19.3 - ~/.nvm/versions/node/v16.19.0/bin/npm
  Browsers:
    Chrome: 108.0.5359.124
    Firefox: 106.0.1
    Safari: 16.0
  npmPackages:
    @storybook/addon-essentials: ^7.0.0-beta.30 => 7.0.0-beta.30 
    @storybook/addon-interactions: ^7.0.0-beta.30 => 7.0.0-beta.30 
    @storybook/addon-links: ^7.0.0-beta.30 => 7.0.0-beta.30 
    @storybook/blocks: ^7.0.0-beta.30 => 7.0.0-beta.30 
    @storybook/react: ^7.0.0-beta.30 => 7.0.0-beta.30 
    @storybook/react-vite: ^7.0.0-beta.30 => 7.0.0-beta.30 
    @storybook/testing-library: ^0.0.13 => 0.0.13
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: https://github.com/storybookjs/storybook/issues/20675

add an esbuild plugin to the builder-manager that injects environment variables

### CODE REVIEW

**Comment by webdeveric** [comment]
@ndelangen The `check-package-manager` cli just ensures the correct package manager is being used in the project. It should skip running if it detects that it is being called from a dependency. Perhaps there is an edge case I missed.

Can you rerun the failed GH Action with debug logging turned on? `check-package-manager` recognizes that debug logging is on and will print out debug messages.

**Comment by IanVS** [comment]
Hah, of course it passed as soon as I enabled debug mode...  Going to try running them all again without, to see if debug actually makes it pass.

**Comment by webdeveric** [comment]
@IanVS I made some fixes to [check-package-manager](https://www.npmjs.com/package/check-package-manager) that resolved the issue on Windows. I'm using `npx` to call the cli so it'll always get the latest version, which is why it is passing now for you.

File: code/lib/builder-manager/package.json [hunk 1]
```diff
@@ -54,6 +54,7 @@
     "ejs": "^3.1.8",
     "esbuild": "^0.16.4",
     "esbuild-plugin-alias": "^0.2.1",
+    "esbuild-plugin-environment": "^0.2.2",
```
Comment (webdeveric) [inline code/lib/builder-manager/package.json:57]: I published version `0.2.3` of this plugin.
If you update to that version, you will get rid of some transitive dependencies.

**Comment by ndelangen** [comment]
We should QA that this feature actually does what we intent it should do.

**Comment by ndelangen** [comment]
@webdeveric I ended up using the `define` property instead, after local testing, it seemed the plugin wasn't replacing references in the code.

I do very much appreciate the support you've given. ❤️

**Comment by webdeveric** [comment]
@ndelangen If you have a reproducible bug, please [open an issue](https://github.com/webdeveric/esbuild-plugin-environment/issues/new) and I'll take a look.

But on first glance, it looks like you were calling `JSON.stringify` when you didn't need to.
My plugin does that for you.

```
environmentPlugin({
  ...stringifyProcessEnvs(envs),
  NODE_ENV: JSON.stringify(process.env.NODE_ENV),
}),
```

Should have been (untested)

```
environmentPlugin(envs),
environmentPlugin(['NODE_ENV']),
```

**Comment by ndelangen** [comment]
Ow, I needed to invoke the plugin twice? once with the object and once with the keys?

I just assumed those were 2 separate use-cases.

**Comment by webdeveric** [comment]
@ndelangen They are different use cases, one for user provided default values, and the other (array of keys) uses empty string as the default.

I was just removing the `JSON.stringify` from your code in my example above.

File: code/ui/.storybook/manager.ts [hunk 2]
```diff
@@ -6,3 +6,6 @@ addons.setConfig({
     renderLabel: ({ name, type }) => (type === 'story' ? name : startCase(name)),
   },
 });
+
+console.log(process.env.STORYBOOK_FOO);
+console.log(process.env);
```
Comment (IanVS) [inline code/ui/.storybook/manager.ts:11]: Did you mean to leave these console.log statements here, @ndelangen ?

Comment (ndelangen) [reply code/ui/.storybook/manager.ts:11]: I did not 😨

**Pull request merged**